server/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java

```diff
@@ -716,10 +719,12 @@ private boolean updateSettings(Settings toApply, Settings.Builder target, Settin
             } else if (get(key) == null) {
                 throw new IllegalArgumentException(type + " setting [" + key + "], not recognized");
             } else if (isDelete == false && canUpdate.test(key)) {
-                validate(key, toApply, false); // we might not have a full picture here do to a dependency validation
-                settingsBuilder.copy(key, toApply);
-                updates.copy(key, toApply);
-                changed = true;
+                changed = hasChanged(toApply, target);
+                if (changed) {
+                    validate(toApply, target.build());
```

Comment:
I think this isn't quite the right approach. As I understand it this change means we're validating the whole settings update here, but by my reading we already validate that later on in the process:

https://github.com/elastic/elasticsearch/blob/4dc3adad5172f1d7c137b983a1715797a9021190/server/src/main/java/org/elasticsearch/action/admin/cluster/settings/SettingsUpdater.java#L94-L95

I think it'd be better to try and do _less_ validation here on settings that have dependent settings. Perhaps the `Setting.Validator` interface should have two `validate()` methods, one for this stage of validation (good for settings with no dependencies, i.e., most of them) and one for the full validation including dependencies.

Reply:
Yes, I see what you mean. I'll try to move the full validation including dependencies in a new API in `Setting.Validator`. I also think it's a better location for this.